warehouse/accounts/auth_policy.py

```diff
@@ -10,15 +10,110 @@
     SessionAuthenticationPolicy as _SessionAuthenticationPolicy,
 )
 
 from warehouse.accounts.interfaces import IUserService
 from warehouse.cache.http import add_vary_callback
 
 
+class AccountTokenAuthenticationPolicy(_CallbackAuthenticationPolicy):
+    def __init__(self, authenticate):
+        self._authenticate = authenticate
+        self.callback = self._auth_callback
+
+        self._routes_allowed = ["forklift.legacy.file_upload"]
+
+    def unauthenticated_userid(self, request):
+        if request.matched_route.name not in self._routes_allowed:
+            return None
+
+        account_token = request.params.get("account_token")
+
+        if account_token is None:
+            return None
+
+        try:
+            macaroon = Macaroon.deserialize(account_token)
+
+            # First, check identifier and location
+            if macaroon.identifier != request.registry.settings["account_token.id"]:
+                return None
+
+            if macaroon.location != "pypi.org":
+                return None
+
+            # Check the macaroon against our configuration
+            verifier = Verifier()
+
+            verifier.third_party_caveat_verifier_delegate = IgnoreThirdPartyCaveats()
+
+            verifier.satisfy_general(self._validate_first_party_caveat)
+
+            verifier.verify(macaroon, request.registry.settings["account_token.secret"])
+
+            # Get id from token
+            account_token_id = None
+            package_list = None
+
+            for each in macaroon.first_party_caveats():
+                caveat = each.to_dict()
+                caveat_parts = caveat["cid"].split(": ")
+                caveat_key = caveat_parts[0]
+                caveat_value = ": ".join(caveat_parts[1:])
+
+                # If caveats are specified multiple times, only trust the
+                # first value we encounter.
+                if caveat_key == "id" and account_token_id is None:
+                    account_token_id = caveat_value
+
+                elif caveat_key == "package_list" and package_list is None:
+                    package_list = caveat_value
+
+            if package_list is not None:
+                request.session["account_token_package_list"] = package_list
+
+            if account_token_id is not None:
+                login_service = request.find_service(IUserService, context=None)
+
+                return login_service.find_userid_by_account_token(account_token_id)
+
+        except (struct.error, MacaroonException):
+            return None
+
+    def remember(self, request, userid, **kw):
+        """A no-op. Let other authenticators handle this."""
+        return []
+
+    def forget(self, request):
+        """ A no-op. Let other authenticators handle this."""
+        return []
+
+    def _validate_first_party_caveat(self, caveat):
+        # Only support 'id' and 'package_list' caveat for now
+        if caveat.split(": ")[0] not in ["id", "package_list"]:
```

Comment:
Yeah, I think this is the most interesting part of the project! There are so many possibilities.

... but I'm also interested in what the minimum requirements are so we can establish this framework (while supporting the future use cases we might want to implement). And of course, whatever is implemented has to be somewhat easy to document and describe.

With that in mind, how about something like:
- `valid_before`: in `'%Y-%m-%d %H:%M:%S%z'` format?
- `valid_after`: same format
    - If both are specified, and the window is more than 2-3 years (taking a cue from TLS certificates), complain? Of course, by default these tokens will not be limited.
- `package_list`: this is where it gets complicated. I think the requirements are:
    - We want to be able to specify multiple packages / versions / actions in one caveat, otherwise that caveat won't validate.
    - We don't want version specification and action to be optional (defaulting to all)
    - We want to be able to further restrict a macaroon with additional caveats later
    - So maybe:
        - A package_list is made of a space-separated list of package_items
        - A package_item has a required package name (canonicalized?)
        - A package_item is optionally followed by `==` and a version string
        - A package_item is optionally followed by a `|` and package actions (like `upload`?)
    - For this diff, we'll just implement the package name and skip the version strings and actions?

Reply:
The ``valid_(before|after)`` syntax seems reasonable. I'm not sure if we want to support all the way out to 2-3 years or not, but ideally that should be a constant defined somewhere we can tweak, so nailing down the exact answer right now isn't super important I don't think.

For ``package_list``, I'm torn. Stepping back a minute, I think these are the sorts of restrictions we want to be able to place on a macaroon:

* Limit the projects for which a macaroon is valid for.
* Limit the releases within that project for which a macaroon is valid for.
* Limit the files (including a hash of the actual file) for which a macroon is valid for.

In the future, we may end up allowing Macaroons to modify other things, like Users? Etc. So I think it's worthwhile to give a little thought to what that would look like. For example, would we just have a ``users_list``` alongside the ``package_list``?

I like the idea of just using the version specifier language to capture the release a thing is valid for, instead of tying that to a separate object. Although I think we should support the full range of version specifiers to make the restrictions syntax more flexible.

One thing to keep in mind with Macaroons, is that one of their key ideas is that given a Macaroon, anyone who possesses it can add additional restrictions going forward before passing that macaroon on to someone else. This makes them very powerful, and can enable a workflow like (just as an example):

1. User grants TravisCI a non-time limited Macaroon for a specific Project.
2. TravisCI, in preparation for a release task, restricts that Macaroon to only be valid for the next 15 minutes before passing it into the environment.
3. Twine takes said macaroon, and further restricts it to only be valid for the specific filename and hash that it's just about to upload.

This is just an example, but it shows off how each layer of the process can further restrict the token, until ultimately the thing that is sent across the wire is only valid for a short period of time, and even if it were stolen, is only usable to upload a specific file anyways.

I lean towards liking the idea of just having a single resource caveat, that has a structure that exposes what kind of resource we're operating on, so my preference would be something like:

``resources: Project("foo") Project("bar>=1.0,<3") File("widget-1.0.tar.gz", hash="sha256:...")``

That makes it a little more complex to parse, but it shouldn't be too bad using a parsing library. I think the API (since the caveat structure is effectively an API) is a bit nicer that way though.

That still leaves the question of whether to have permissions be tied to a specific resource, or be a global list (or even both).  I'm not sure here! If we're tying them to per resource, it'd just be a "kwarg" to the above structure. If we're making them global they'd just be their own list.

Actually, as I'm typing this it just occurred to me, there's no reason that the actual caveat language needs to be human friendly, it can be machine friendly and the tooling can provide the human friendly layer on top of it. So one possible option here is instead of defining a textual serialization, we can just define a structure, like (excuse the typing syntax):

```python
import enum

from typing import List, Mapping, Union, Optional
from mypy_extensions import TypedDict

class ResourceKind(enum.Enum):
    Project = "Project"
    File = "File"

class Resource(TypedDict, total=True):
    kind: ResourceKind

# Due to limitation in MyPy, we have to put mandatory and
# optional keys in separate classes.
class _Project(Resource, total=True):
    name: str

# All of these keys are optional.
class Project(_Project, total=False):
    version_specifier: str

class _File(Resource, total=True):
    filename: str

class File(_File, total=False):
    hashes = TypedDict("Hashes", {"sha256": str}, total=False)

Resource = Union[Project, File]

ResourcesCaveat = List[Resource]
```

This would mean that basically the caveat would end up looking like

```python
resources = [
    {"kind": "Project", "name": "foo"},
    {"kind": "Project", "name": "bar", "verson_specifier": ">=1.0,<3"},
    {"kind": "File", "filename": "widget-1.0.tar.gz", "hashes": {"sha256": "..."}},
]

caveat = f"resources: {serialize(resources)}"
```

The the question becomes what serialization language do we support. JSON is an easy one because it's built into Python, ubiquitous, and works the same on 2.x and 3.x. The biggest downside to JSON is that it's not really very compact. Another option I like is [msgpack](https://msgpack.org/), which is a binary protocol (so we'd have to make sure that Macaroons doesn't choke if you put binary blobs into the caveat language) but it's much more efficient and compact than JSON.

One nice thing about msgpack is that it even supports the idea of extension types, so we could even ditch the ``kind`` field, and just register each of the "types" of resources as an extension type in MsgPack. That would make the wire protocol a bit more compact, but would make serializing/deserializing a bit more complex and would tie us to msgpack. I'm not sure if the tradeoff is worth it, but it might be, it would depend on how much more compact it made the wire protocol.

If we did go with a binary that would open up the question of if it would make sense to apply compression to the serialized binary, and if that would make the serialized caveat shorter or not (and if it did, shorter enough to make a difference). One of the main downsides with Macaroons is that all of the caveats are embedded into the macaroon itself, so each one you add, increases the length of the macaroon, so compression could help combat that.

Taking this idea even further... I wonder if it would make sense to just define *all* of our caveats in terms of a single structure like that, so that instead of having a ``resources`` caveat, and a ``not_before`` caveat, and a ``not_after`` caveat, etc etc, we instead have a single caveat without an identifier (so we don't need to try and do string splitting or whatever) that has a structure like (extending the structure from before):

```python

class Expiration(TypedDict, total=True):
    not_before: datetime.datetime
    not_after: datetime.datetime

class Caveat(TypedDict, total=False):
    expiration: Expiration
    resources: ResourcesCaveat

```

Which would then ultimately give us a structure like:

```python
caveat = {
    "expiration": {
        "not_before": datetime.datetime(...),
        "not_after": datetime.datetime(...),
    },
    "resources": [
         {"kind": "Project", "name": "foo"},
         {"kind": "Project", "name": "bar", "verson_specifier": ">=1.0,<3"},
         {"kind": "File", "filename": "widget-1.0.tar.gz", "hashes": {"sha256": "..."}},
    ],
}

caveat = serialize(caveat)
```

Then adding more caveats is just adding more keys to the ``Caveat`` type, it would obviously be important that we reject any caveat that has keys that we don't understand. 

With that being said, I'm fine with deferring some of the caveat features into a later PR. Ultimately one could argue that we could even defer *all* of them into a later PR and let the first PR focus on getting the Macaroon framework in place (since without caveats, Macaroons are not worse than the status quo, and are a little bit better in that you can have multiple of them and revoke them independently, etc).